### PR TITLE
Update counts to handle floats/NaN values from hardware

### PR DESCRIPTION
**Pull request by lillian542**
The QuEra hardware sometimes has failed measurements, which return as NaN. Because of this, all samples from the device are returned with type `float`, and some of them are `np.NaN`. These were not handled well in counts. On the `_default_qubit` device, nothing was converted to integers, so it did this:

```
{'0.00.00.0': 149,
'0.00.01.0': 281, 
'0.00.0nan': 2,
'0.01.00.0': 277,
'0.01.01.0': 4,
'0.0nan0.0': 4,
'1.00.00.0': 268,
'1.00.0nan': 2,
'1.01.00.0': 5,
'nan0.00.0': 8}
```

We don't like that. Now they are converted into integers if there is no observable, just like in `CountsMP`.

Also added a line both there and in `CountsMP` to avoid converting the `NaN` values to 0 and implying they were ground state measurements - instead, any sample where at least one wire failed is disregarded when totalling the number of counts. 

Has been tested with the QuEra hardware, and now returns nicely:

```
>>> res = circuit(params)
>>> res
{'000': 230, '001': 251, '010': 252, '011': 1, '100': 249, '110': 4}
```

with 13 disregarded failed measurements:
```
>>> sum(count for count in res.values())
987
```

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-48588]

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -1437,6 +1437,8 @@ def circuit(x):
 
         if isinstance(obs, CountsMP):
             # convert samples and outcomes (if using) from arrays to str for dict keys
+            samples = np.array([sample for sample in samples if not np.any(np.isnan(sample))])
```
Comment (albi3ro) [inline pennylane/_qubit_device.py:1440]: Should we also have this fix in `CountsMP.process_samples`?

Comment (lillian542) [reply pennylane/_qubit_device.py:1440]: I think it's a good idea, I'll add that and a couple of tests

**Review by albi3ro — APPROVED** [review]
changelog?

Thanks for this fix 🎉

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (mudit2812) [inline pennylane/_qubit_device.py:1440]: Are there any instances where we have to worry about other interfaces? I added `qml.math.isnan` and `qml.math.any` as part of my postselection PR.

Comment (mudit2812) [reply pennylane/_qubit_device.py:1440]: Ah nevermind QubitDevice just uses numpy.

**Review by mudit2812 — APPROVED** [review]

**Pull request merged**